network/src/peers/peer/outbound_handler.rs

```diff
@@ -100,21 +109,45 @@ impl Peer {
     ) -> Result<PeerResponse, NetworkError> {
         match message {
             PeerAction::Disconnect => Ok(PeerResponse::Disconnect),
+            PeerAction::SendBlock(block, height) => {
+                // check if they sent us the block recently
+                let block = block.serialize();
+                let message = Payload::Block(block, Some(height));
+
+                self.queued_outbound_message_count.fetch_sub(1, Ordering::SeqCst);
+                metrics::decrement_gauge!(OUTBOUND, 1.0);
+
+                if self.block_received_cache.contains(&message) {
+                    metrics::increment_counter!(metrics::outbound::ALL_CACHE_HITS);
+                    return Ok(PeerResponse::None);
+                }
+
+                network.write_payload(&message).await.map_err(|e| {
+                    metrics::increment_counter!(metrics::outbound::ALL_FAILURES);
+                    e
+                })?;
+
+                metrics::increment_counter!(metrics::outbound::ALL_SUCCESSES);
+
+                debug!("Sent a '{}' message to {}", &message, self.address);
+                Ok(PeerResponse::None)
+            }
             PeerAction::Send(message) => {
                 if matches!(message, Payload::Ping(_)) {
                     self.quality.expecting_pong = true;
                     self.quality.last_ping_sent = Some(Instant::now());
                 }
+
+                self.queued_outbound_message_count.fetch_sub(1, Ordering::SeqCst);
+                metrics::decrement_gauge!(OUTBOUND, 1.0);
+
```

Comment:
why was the metric change moved here, before the outbound message has been written to the socket? I think it made more sense for it to be subtracted only once the message is physically out

Reply:
it's so that cache hits and failures also decrement the outbound metric.

Reply:
good point wrt `SendBlock`, though to be most correct in terms of the number of items waiting to be written to the socket, we should just add an extra `decrement_gauge` in case of a cache hit and place the existing ones after the writes

Reply:
considering the difference is off by one, i'm inclined to say screw in in favor of cleanliness here. let me know if you disagree.